### PR TITLE
Instead of copying the entire installer, download drivers on demand

### PR DESCRIPTION
**Pull request by zanderz**
The current hack of copying the entire installer is fragile and skirts some normal practices, and is bound to cause some problems sooner or later.

I also figured out that the regular installer cache can be made to contain these components after all, but we can't use that technique because per-machine components require UAC to even be placed in the cache, breaking silent upgrade for people who haven't installed drivers.

@keybase/react-hackers @maxtaco

### CODE REVIEW

File: packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs [hunk 1]
```diff
@@ -224,25 +226,27 @@
...
-      <MsiPackage SourceFile="$(var.DOKAN_PATH)\dokan_wix\bin\x86\$(var.Configuration)\Dokan_x86.msi"
+      <MsiPackage Name="Dokan_x86.msi"
+                  SourceFile="$(var.DOKAN_PATH)\dokan_wix\bin\x86\$(var.Configuration)\Dokan_x86.msi"
                   InstallCondition="NOT VersionNT64 AND (driver OR DokanUninstallString)"
-                  Compressed="yes"
+                  Compressed="no"
                   Visible="yes"
                   Permanent="yes"
-		  Cache="yes"
+		  DownloadUrl="http://s3.amazonaws.com/prerelease.keybase.io/windows-support/dokan-dev/build84/dokan_wix/bin/x86/$(var.Configuration)/Dokan_x86.msi"
```
Comment (maxtaco) [inline packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs]: this really ought to be http, and some DNS name that's ours.

File: packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs [hunk 2]
```diff
@@ -224,25 +226,27 @@
...
-      <MsiPackage SourceFile="$(var.DOKAN_PATH)\dokan_wix\bin\x64\$(var.Configuration)\Dokan_x64.msi"
+      <MsiPackage Name="Dokan_x64.msi"
+                  SourceFile="$(var.DOKAN_PATH)\dokan_wix\bin\x64\$(var.Configuration)\Dokan_x64.msi"
                   InstallCondition="VersionNT64 AND (driver OR DokanUninstallString)"
-                  Compressed="yes"
+                  Compressed="no"
                   Visible="yes"
                   Permanent="yes"
-		  Cache="yes"
+		  DownloadUrl="http://s3.amazonaws.com/prerelease.keybase.io/windows-support/dokan-dev/build84/dokan_wix/bin/x64/$(var.Configuration)/Dokan_x64.msi"
```
Comment (maxtaco) [inline packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs]: ditto as above

**Comment by zanderz** [comment]
It took a long time to figure out, but at build time the local file pointed to by `SourceFile` has its hash baked in to the installer, and if at install time the downloaded one doesn't match, it errors out. You may notice that the build process downloads these from s3 if they aren't available locally already, but we could change that part.

Again, this is just a suggestion, we don't have to do this, but I thought it important enough to spend my 2-pointer on.

**Comment by maxtaco** [comment]
Ok, that's a good plan. But you should still download over HTTPS. And the dokanclean process doesn't seem to have this security mechanism built it. It seems to just exec whatever it downloads from S3 (over HTTP).

**Comment by zanderz** [comment]
Dokanclean gets downloaded if not already locally copied, just as a convenience to avoid another build machine setup step. I added a hash check here but we could just skip that part of the build and do it manually if you prefer. The installer will still enforce at runtime that the one it downloads matches the one present when the installer was built.

File: packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs [hunk 3]
```diff
@@ -98,7 +98,7 @@
         PerMachine="yes"
         Permanent="yes"
         Compressed="no"
-        DownloadUrl="http://s3.amazonaws.com/prerelease.keybase.io/windows-support/dokan-dev/build84/dokanclean.exe"
+        DownloadUrl="https://s3.amazonaws.com/prerelease.keybase.io/windows-support/dokan-dev/build84/dokanclean.exe"
```
Comment (maxtaco) [inline packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs]: can you use `https://prerelase.keybase.io` instead?

File: packaging/windows/doinstaller_wix.cmd [hunk 4]
```diff
@@ -127,6 +126,10 @@ IF %ERRORLEVEL% NEQ 0 (
   EXIT /B 1
 )
 
+:: Sanity check the hash of dokanclean downloaded during installer build
+for /f "usebackq tokens=2*" %%i in (`powershell Get-FileHash -Algorithm sha1 %GOPATH%\src\github.com\keybase\client\packaging\windows\WIXInstallers\KeybaseBundle\Redist\dokanclean.exe`) do set DOKANCLEANHASH=%%i
```
Comment (maxtaco) [inline packaging/windows/doinstaller_wix.cmd]: Please use use `sha2`.

**Comment by maxtaco** [comment]
Got it, can you point me to the place in the code where the hashes are checked on the other side?

File: packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs [hunk 5]
```diff
@@ -92,12 +92,14 @@
...
       <ExePackage
-        SourceFile="$(env.GOPATH)\src\github.com\keybase\client\go\tools\dokanclean\dokanclean.exe"
+        SourceFile="Redist\dokanclean.exe"
         DetectCondition="DokanUninstallString OR NOT driver"
         Description="Remove previous drivers"
         PerMachine="yes"
         Permanent="yes"
-	Cache="yes">
+        Compressed="no"
+        DownloadUrl="https://s3.amazonaws.com/prerelease.keybase.io/windows-support/dokan-dev/build84/dokanclean.exe"
```
Comment (maxtaco) [inline packaging/windows/WIXInstallers/KeybaseBundle/Bundle.wxs]: please use `https://prerelease.keybase.io`

**Comment by zanderz** [comment]
Install time verification of hashes is in the Wix Burn source code:
https://github.com/wixtoolset/wix3/blob/c2b0b3ff94acda436ea05dce8ebc7e659cf1fc24/src/burn/engine/cache.cpp#L1825

**Comment by maxtaco** [comment]
I guess that means we can't use sha2!

**Comment by maxtaco** [comment]
Where do you specify the hashes for the various downloaded packages?

**Comment by zanderz** [comment]
You don't specify the hash, you specify a local copy, kind of a reference copy, which it hashes itself at build time, then bakes the hashes into the installer. Note I just removed the downloading of dokanclean.exe as a build step; we'll set that up manually on each build machine.

It is true that we have little control over this in the current toolkit, but I guess we could add a different hashing mechanism and build our own WIX and/or submit a PR. Also note, our installer has been downloading .NET4 this way the whole time.

**Comment by maxtaco** [comment]
Understood, thank you for the explanation.  Can you test this by changing the image on the server after it's built and making sure the hash check fails?

**Comment by zanderz** [comment]
yep, already checked that, but can do again

**Comment by maxtaco** [comment]
ok, thanks!

**Comment by zanderz** [comment]
Here you go (changed the s3 file after building):
image
the log says which file, and what the expected + found hashes were

**Comment by maxtaco** [comment]
cool. Gotta love those windows error codes! `0x80091007` !

**Pull request merged**